### DISCUSSION TITLE
Multiline HTML comments and evaluation of chunks

### DISCUSSION THREAD

**Discussion by boshek** [quarto-dev/quarto-cli#9976] [Q&A] [answered]
### Description

I see discussions on commenting [here](https://github.com/quarto-dev/quarto-cli/discussions/3330) and [here](https://github.com/quarto-dev/quarto-cli/issues/696#issuecomment-1256636144) but I haven't seen anything the outlines this specific issue. Apologies if I have missed it.

Consider this quarto doc called `comment-reprex.qmd`:

````qmd
---
format: 
  html: 
    self-contained: true
---

```{r}
head(iris)
```

{{< include child-doc.qmd >}}
````

and this child doc called `child-doc.qmd`:
````qmd
```{r}
head(quakes)
```
````

Rendering with `quarto render` produces an html that looks like this:
![image](https://github.com/quarto-dev/quarto-cli/assets/18472598/dc7f53b3-2f14-4257-8ccb-69af2249f1d1)

Great. That works fine. But I am seeing some (what I think to be) strange behaviour with multiline line comments. For example, a multiline comment for the R chunk removes it from the output but the chunk is still evaluated (checked with `grep "Sepal" comment-reprex.html`):

````qmd
---
format: 
  html: 
    self-contained: true
---

<!-- 
```{r}
head(iris)
```
 -->

{{< include child-doc.qmd >}}
````

The same scenario happens when I multiline comment out the `include` line (checked with `grep "stations" comment-reprex.html`):

````qmd
---
format: 
  html: 
    self-contained: true
---


```{r}
head(iris)
```

<!-- 
{{< include child-doc.qmd >}}
 -->
````

However if one simply does a single line comment  the chunks are _not_ evaluated (again checked with `grep "stations" comment-reprex.html`):
````qmd
---
format: 
  html: 
    self-contained: true
---


```{r}
head(iris)
```


<!-- {{< include child-doc.qmd >}} -->
````

Is this a known pattern? I don't see anything mentioned in the [commenting section](https://quarto.org/docs/visual-editor/content.html#commenting) so I am raising this here. If this is an unavoidable issue, for now would you consider a PR to add some text around multiline comments in that section? 

Tested on
- macOS
- quarto version 1.5.27

**Comment by cscheid** [comment]
HTML comments are not supported in Quarto (or Pandoc).

**Reply by cscheid** [reply, marked as answer]
```
<!-- {{< include child-doc.qmd >}} -->
```

Note the Important callout in [our documentation](https://quarto.org/docs/authoring/includes.html):

> Important: Include shortcodes need to appear by themselves in a line, and they need to be surrounded by empty lines. This means that you cannot use an include shortcode inside markdown syntax (such as an item in a bulleted list).

**Reply by boshek** [reply]
Thanks this is helpful. But maybe I missing something with this:

> HTML comments are not supported in Quarto (or Pandoc).

Because `<!-- {{< include child-doc.qmd >}} -->` _does_ work at least from a user facing perspective.

**Reply by cderv** [reply]
> For example, a multiline comment for the R chunk removes it from the output but the chunk is still evaluated (checked with grep "Sepal" comment-reprex.html):

This is an illustration of HTML document are not supported to ignore content. In your case, the knitr engine will still execute and run the chunk because it is seen as a cell and the HTML comment means nothing. 

For knitr, there is the specific comment engine : https://yihui.org/en/2022/01/knitr-news/#the-new-engines-comment-verbatim-and-embed that quarto inherits from using **knitr**

**Reply by boshek** [reply]
That is a really helpful link thank you. The thing that I am still confused about then is why this works: 

```
<!-- {{< include child-doc.qmd >}} -->
```

The vscode extension will [happily](https://github.com/quarto-dev/quarto/blob/06678d55143a7d6de6c7f0232db0c5dcf8392ca6/apps/vscode/language-configuration.json#L4) comment out html code which _does_ suggest that comments should work?

**Reply by cscheid** [reply]
Yuck, I don't think that should be there.

**Reply by boshek** [reply]
Sorry just re-upping this as I think it is relevant. Anyone have any thoughts on:

>  `<!-- {{< include child-doc.qmd >}} -->` does work at least from a user facing perspective.

From the discussion above, should I take away that this _isn't_ supposed to work?

**Reply by cscheid** [reply]
What is the `this` that you're asking about? An include shortcode needs to be on a line by itself, per the documentation. HTML comments "don't do anything" in Markdown; they're interpreted as an HTML RawBlock.

This gets passed through to the HTML format but nowhere else.

<img width="698" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/84620905-2fbd-48c2-96c8-3d8f18bd56f1">

And so does that line:

<img width="340" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/aebd799e-8d01-4b65-89f6-0829b4d9739d">

I'm not sure what you're trying to say.

**Reply by mcanouil** [reply]
@boshek dis you miss <https://github.com/quarto-dev/quarto-cli/discussions/9976#discussioncomment-9754291>?

**Reply by boshek** [reply]
Sorry for being obtuse. Functionally if I do this:

````qmd
---
format: 
  html: 
    self-contained: true
---


<!-- {{< include child-doc.qmd >}} -->
````

It acts like a code comment as any content in `child-doc.qmd` is _not_ included in the final HTML.

**Reply by boshek** [reply]
I see better now.... so the inclusion of the `<!--` character in front of the `include` means that it becomes a non-valid `include` syntax. But I think the point stands here that _functionally_ it acts like a code comment that works whereas the multiline comment does not. I appreciate you leading me through this.

**Reply by mcanouil** [reply]
That's only because this blocks parsing of the shortcode.
As stated, the include shortcode requires to be on its own line.

```
Whatever {{< include child-doc.qmd >}}
```

Will also make the include shortcode be not evaluated.

Regarding code cells, knitr as Quarto/Pandoc does not care about HTML comments syntax. It will evaluate everything unless `eval: false`.

Note that the correct/current (for almost 2 years now) option is `embed-resources` not `self-contained`. It only works because of backward compatibility, but you should not rely on that for current and future document you create.

**Reply by boshek** [reply]
> Note that the correct/current (for almost 2 years now) option is embed-resources not self-contained. It only works because of backward compatibility, but you should not rely on that for current and future document you create.

Thank you!

> That's only because this blocks parsing of the shortcode.

I get that. 100%. Totally appreciate you folks spelling that out for me. The thing that I am trying to express is that _functionally_ without a user knowing all this context, it will act like a code comment, potentially confusing a user (like myself). Maybe there is not solution to this but I am just trying to express that this is a subtle difference that may trip others up as well.

**Reply by mcanouil** [reply]
Everything is basically discussed in https://github.com/quarto-dev/quarto-cli/discussions/3330

Even if some kind of commenting system is implemented (which html comment is not), `knitr` will still evaluate any code cells it finds, so the issue has many levels.
